src/Microsoft.Build.Tasks.Git/GitDataReader/GitRepository.cs

```diff
@@ -209,8 +209,19 @@ public static GitRepository OpenRepository(GitRepositoryLocation location, GitEn
         /// <returns>Null if the HEAD tip reference can't be resolved.</returns>
         internal string? ReadSubmoduleHeadCommitSha(string submoduleWorkingDirectoryFullPath)
         {
-            var gitDirectory = ReadDotGitFile(Path.Combine(submoduleWorkingDirectoryFullPath, GitDirName));
-            if (!IsGitDirectory(gitDirectory, out var commonDirectory))
+            // Submodules don't usually have their own .git directories but this is still legal.
+            // This can occur with older versions of Git or other tools, or when a user clones one
+            // repo into another's source tree (but it was not yet registered as a submodule).
+            // See https://git-scm.com/docs/gitsubmodules#_forms for more details.
+            // Handle this case first since the other case throws.
+            var dotGitPath = Path.Combine(submoduleWorkingDirectoryFullPath, GitDirName);
+
+            var gitDirectory =
+                Directory.Exists(dotGitPath) ? dotGitPath :
+                File.Exists(dotGitPath) ? ReadDotGitFile(dotGitPath) :
```

Comment:
A test is failing - looking at it I think we should keep throwing if the file can't be read, rather then returning null. 

```C#
var gitDirectory = Directory.Exists(dotGitPath) ? dotGitPath : ReadDotGitFile(dotGitPath);

if (!IsGitDirectory(gitDirectory, out var commonDirectory))
```

Reply:
With this change you'd still need to update the `Submodules_Errors` test - remove case S6:
```
[submodule ""S6""]             # sub6/.git is a directory, but should be a file
  path = sub6
  url = http://github.com
````

Reply:
Or rather update it - since in this case the directory is not a valid git directory (it's empty) so failure is expected of some sort